File: pyscripts/filter_og2genes.py
#!/bin/env python3

import argparse
import os
import gzip
import sys

og_ids    = set()
spp_ids   = set()
og2genes  = '' # input files
og_ids_f  = ''
spp_ids_f = '' 

def get_arguments() -> int:
    """get the arguments"""

    global og2genes, og_ids_f, spp_ids_f

    d = "Collect the gene records from odb12v2_OG2genes.tab.gz using OG and species IDs"

    parser = argparse.ArgumentParser(description = d)
    parser.add_argument("-o", "--ogs", help="odb12v2_OG2genes.tab.gz file", required=True, type=str)
    parser.add_argument("-g", "--groups", help="single columns list of OG IDs", required=True, type=str)
    parser.add_argument("-s", "--spp", help="single columns list of species IDs", required=True, type=str)
    args      = parser.parse_args()
    og2genes  = args.ogs
    og_ids_f  = args.groups
    spp_ids_f = args.spp

    assert os.path.isfile(og2genes), f"Could not locate file {og2genes}"
    assert os.path.isfile(og_ids_f), f"Could not locate file {og_ids_f}"
    assert os.path.isfile(spp_ids_f), f"Could not locate file {spp_ids_f}"

    return 0

def get_og_ids() -> int:
    """load the OG ids into memory"""

    global og_ids, og_ids_f

    fh = gzip.open(og_ids_f, "rt") if og_ids_f.endswith(".gz") else open(og_ids_f, 'r')

    for line in fh:
        if (len(line) == 0 or line[0] == '#'):
            continue
        og_id = line.strip()
        og_ids.add(og_id)

    fh.close()

    print(f"Loaded {len(og_ids)} target groups")

    return 0

def get_spp_ids() -> int:
    """load the species ids into memory"""

    global spp_ids, spp_ids_f

    fh = gzip.open(spp_ids_f, "rt") if spp_ids_f.endswith(".gz") else open(spp_ids_f, 'r')

    for line in fh:
        if (len(line) == 0 or line[0] == '#'):
            continue
        spp_id = line.strip()
        spp_ids.add(spp_id)

    fh.close()

    print(f"Loaded {len(spp_ids)} target species")

    return 0

def make_output_name() -> str:
    """helper function to create the output name"""

    global og_ids, spp_ids

    out = f"filtered.{len(og_ids)}.OGs.{len(spp_ids)}.spp.odb12v2_OG2genes.tab.gz"

    return out

def get_og_genes() -> int:
    """the work horse of this application"""

    global og_ids, spp_ids, og2genes

    fh  = gzip.open(og2genes, "rt") if og2genes.endswith(".gz") else open(og2genes, 'r')
    ofh = gzip.open(make_output_name(), "wt")
    tot = 0

    for line in fh:
        if (len(line) == 0 or line[0] == '#'):
            continue

        cnt = 0
        for i in range(len(line)):
            if (line[i] == '\t' or line[i] == ':'):
                cnt += 1

        # safety check
        if (cnt != 2):
            msg = f"Invalid line found:\n{line}"
            sys.exit(msg)

        fields = line.split('\t')
        og_id  = fields[0]

        # ogroup checkpoint
        if (og_id not in og_ids):
            continue

        # species checkpoint
        idx = fields[1].find(':')
        spp_id = fields[1][:idx]
        if (spp_id not in spp_ids):
            continue

        ofh.write(line)
        tot += 1

    fh.close()
    ofh.close()

    print(f"Retained {tot} records")

    return 0

def main() -> int:
    """entry point to this little subprogram"""

    # get arguments
    get_arguments()

    # load the target OGs into memory
    get_og_ids()

    # load the species IDs into memory
    get_spp_ids()

    # start processing
    get_og_genes()

    return 0

if __name__ == "__main__":
    main()

File: cpp.files/filter_og2genes.cpp
//
// parse a bam file to find reads bridging 2 seperate contigs
//

#include <iostream>
#include <fstream>
#include <algorithm>
#include <sys/stat.h>
#include <string>
#include <unordered_set>
#include <vector>
#include <cstdio>
#include <cctype>
#include <zlib.h>

// make sure to compile with -lz

using std::string;
using std::unordered_set;
using std::vector;
using std::fstream;
using std::iostream;
using std::cout;
using std::cerr;
using std::to_string;

typedef unsigned int uint;

int
parse_tabular(string &line, vector<string> &parts){

    int start  = 0, end = 0;
    int colons = 0;

    while (end < line.size()){
        if (line[end] == '\t'){
            parts.emplace_back(line.substr(start, end - start));
            start = end + 1;
        }
        else if (line[end] == ':'){
            colons++;
        }
        end++;
    }

    if (start < line.size()){
        parts.emplace_back(line.substr(start));
    }

    if (colons != 1 || parts.size() != 2){
        cerr << "Invalid line found\n" << line;
        exit(1);
    }

    return 0;
}

bool
file_exists(const string &path){
    struct stat buffer;
    return stat(path.c_str(), &buffer) == 0;
}



int parse_targets(string &infile, unordered_set<string> &targets, bool is_og){
    //
    // parse the targets file and add the records to the target set
    //

    fstream file(infile);

    if (!file.is_open()){
        cerr << "Error: Could not open " << infile << '\n';
        exit(1);
    }

    string target;
    while (std::getline(file, target)){
        targets.insert(target);
    }

    file.close();

    string t = is_og ? "groups" : "species";

    cerr << "Loaded " << targets.size() << " target " << t << '\n';

    return 0;
}

int
filter_orthogroups(string og2genes_file, unordered_set<string> &og_ids, unordered_set<string> &spp_ids){
    //
    // this is the main work horse for this function
    //

    int spp_count = spp_ids.size(), og_count = og_ids.size();

    gzFile fh = gzopen(og2genes_file.c_str(), "rb");

    if (fh == NULL){
        cerr << "Error: could not open " << og2genes_file << '\n';
        exit(1);
    }

    string outname = "filtered." + to_string(og_count) + ".OGs." + to_string(spp_count) +
                     ".spp.odb12v2_OG2genes.tab.gz";

    gzFile ofh = gzopen(outname.c_str(), "wb");
    
    if (ofh == NULL){
        cerr << "Error: could not open " << outname << '\n';
        exit(1);
    }

    // create the buffer
    const int buff_size = 8192;
    char buffer[buff_size];
    long tot = 0; // keep count

    // start parsing
    vector<string> parts;
    while (gzgets(fh, buffer, buff_size) != NULL){
        string line(buffer);

        if (!line.empty() && line.back() == '\n'){
            // strip new line char
            line.pop_back();
        }
        if (line.empty()){
            continue;
        }
        // clear then fill
        parts.clear();
        parse_tabular(line, parts);

        const string og_id = parts[0];

        // first checkpoint
        if (og_ids.count(og_id) == 0){
            continue;
        }

        // second checkpoint
        const string spp_id = parts[1].substr(0, parts[1].find(':'));

        if (spp_ids.count(spp_id) == 0){
            continue;
        }

        // write this record
        gzprintf(ofh, "%s\t%s\n", og_id.c_str(), parts[1].c_str());
        tot++;

    }

    // close out
    gzclose(fh);
    gzclose(ofh);

    cerr << "Retained " << tot << " records\n";

    return 0;
}

void
help(){
    //
    // help message
    //
    cerr << "./filter_og2_genes \n"
         << "-s species_list [single column list of species IDs]\n"
         << "-g og_list [single column list of OG to select]\n"
         << "-o [odb12v2_OG2genes.tab.gz file]\n";
    exit(0);

}

int
main(int argc, char *args[]){

    string og_file, spp_file, og2genes_file;
    unordered_set<string> spp_ids, og_ids;

    for (int i = 1; i < argc; i++){
        string arg = args[i];
        if (arg == "-s" && i + 1 < argc){
            spp_file = string(args[i + 1]);
        }
        else if (arg == "-g" && i + 1 < argc){
            og_file = string(args[i + 1]);
        }
        else if (arg == "-o" && i + 1 < argc){
            og2genes_file = string(args[i + 1]);
        }
        else if (arg == "-h"){
            help();
        }
    }

    if (og_file.size() == 0 || !file_exists(og_file)){
        cerr << "Unable to use orthogroup list " << og_file << '\n';
        exit(1);
    }
    else if (spp_file.size() == 0 || !file_exists(spp_file)){
        cerr << "Unable to use species list " << spp_file << '\n';
        exit(1);
    }
    else if (og2genes_file.size() == 0 || !file_exists(og2genes_file)){
        cerr << "Unable to use odb12v2_OG2genes.tab.gz file provided as " << og2genes_file << '\n';
        exit(1);
    }

    //
    // step 1: load all the targets
    //
    parse_targets(og_file, og_ids, true);
    parse_targets(spp_file, spp_ids, false);

    //
    // step 2: write the output file as we go
    //
    filter_orthogroups(og2genes_file, og_ids, spp_ids);

    return 0;
}